### DISCUSSION TITLE
Adding per-page dynamic data to Starlight `<head><meta>` tags

### DISCUSSION THREAD

**Discussion by donmckenna** [withastro/starlight#1106] [Q&A] [answered]
### What version of `starlight` are you using?

^0.13.0

### What is your idea?

If this functionality exists, I can't seen to find a reference to it.

I'm using Starlight for a new docs site.

I need a dynamic `meta` `og:image` url, using props based on the page requesting the image.
This is simple in a basic Astro project if I were using my own `Head` / `BaseHead` `.astro` layout component by grabbing page/url props from the `Astro` global.

But afaik Starlight doesn't expose this component to edit, and I don't want to config-override the entire `Head` component just for a few meta tags.
Instead of overriding the `Head` component, the docs recommend using the `head` option in `astro.config`, but afaik page-specific props aren't accessible in `astro.config`, so I can't make these options dynamic.

I could explicitly define `head/meta` attributes in the frontmatter of each `.md` post, but that is cumbersome and defeats the purpose of the dynamic / automation functionality I'm trying to achieve.

What is the best way to achieve a dynamic `meta og:image` url, based on page-specific attributes/route/params, in Starlight?

### Why is this feature necessary?

Would offer a less cumbersome experience getting page-specific, dynamic data into `<head>` tags in Starlight specifically.

### Participation

- [ ] I am willing to submit a pull request for this feature.

**Comment by delucis** [comment]
For OG image specifically, overriding the `Head` component is probably fine! Especially if you follow the [“Reuse a built-in component”](https://starlight.astro.build/guides/overriding-components/#reuse-a-built-in-component) pattern to preserve the default `<head>` content.

The main issue with overriding `Head` is that there is a risk of creating duplicate content, for example when overriding like this:

```astro
---
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/components/Head.astro';
---

<Default {...Astro.props}><slot /></Default>
<title>My custom title</title>
```

You end up with two `<title>` tags because the default component renders one and then the override adds another. When you, use the `head` configuration options in `astro.config.mjs` or in frontmatter, Starlight is able to deduplicate `title` and other tags that shouldn’t be repeated.

Because Starlight does not inject any `og:image` tags for you, this isn’t an issue in this case though, so it should be safe.

That said, Open Graph image support is something we’d like to improve, so I realise this isn’t necessarily an ideal answer just yet.

**Reply by SamuelLHuber** [reply]
love that! thanks for the amazing explainer. How can I override the custom Head in Frontmatter?

say I overrite the Head component with content:

```
---
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/components/Head.astro';
---

<Default {...Astro.props}><slot /></Default>
<meta property="fc:frame" content="vNext" />
<meta property="fc:image" content="https://webpage.com/custom.png" />
```

and now in my markdown file frontmatter I want override head <meta> tags for the property fc:frame and fc:image

```
---
title: my custom markdown html <head>
head:
  - tag: meta
    attrs: { 
      property: 'fc:frame',
      content: vNext
    }
  - tag: meta
    attrs: { 
      property: 'fc:image',
      content: https://webpage.com/custom.png
    }
---

**markdown content**
```

In my testing the Markdown Frontmatter Head is just added to the page which means I need to manually write out the head tags in each frontmatter

**Reply by HiDeoo** [reply]
> How can I override the custom Head in Frontmatter?

Your `<Head>` component override can [access](https://starlight.astro.build/reference/overrides/#component-props) a standard `Astro.props` object that contains information about the current page including its frontmatter values. Considering this, you could [extend](https://starlight.astro.build/reference/frontmatter/#extend) the frontmatter schema to include an optional `fcFrame` field:

```ts
// src/content/config.ts
import { defineCollection, z } from 'astro:content';
import { docsSchema } from '@astrojs/starlight/schema';

export const collections = {
  docs: defineCollection({
    schema: docsSchema({
      extend: z.object({
        // Add a new optional field to the schema.
        fcFrame: z.string().optional(),
      }),
    }),
  }),
};
```

This means that you can now add or not a `fcFrame` field to the frontmatter of any content file which can be used in your component override to either render the custom value from the frontmatter or a default value:

```astro
---
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/components/Head.astro';
---

<Default {...Astro.props}><slot /></Default>
<!-- Either render the value specified in the frontmatter or a default one. -->
<meta property="fc:frame" content={Astro.props.entry.data?.fcFrame ?? "vNext"} />
// …
```

The same could be done for the `fc:image` too.

**Reply by SamuelLHuber** [reply]
lovely thank you 😍 much appreciated

**Comment by HiDeoo** [comment, marked as answer]
Starting with [Starlight 0.35.0](https://github.com/withastro/starlight/releases/tag/%40astrojs%2Fstarlight%400.35.0), [route data](https://starlight.astro.build/guides/route-data/) now include the [`head` property](https://starlight.astro.build/reference/route-data/#head) which contains an array of all tags to include in the <head> of the current page.

Such data can be customized using a [route data middleware](https://starlight.astro.build/guides/route-data/#how-to-customize-route-data). For example, to add some meta tags for Open Graph images, this would look like the following:

```js
// src/routeData.ts
import { defineRouteMiddleware } from "@astrojs/starlight/route-data";

export const onRequest = defineRouteMiddleware((context) => {
  // Get the URL of the Open Graph image to use.
  const ogImageUrl = new URL(`/og/my-og-image.png`, context.site);

  // Get the array of all tags to include in the `<head>` of the current page.
  const { head } = context.locals.starlightRoute;

  // Add the `<meta/>` tags for the Open Graph images.
  head.push({
    tag: "meta",
    attrs: { property: "og:image", content: ogImageUrl.href },
  });
  head.push({
    tag: "meta",
    attrs: { name: "twitter:image", content: ogImageUrl.href },
  });

  // Add the `fc:frame` meta tag.
  head.push({
    tag: "meta",
    attrs: { property: "fc:frame", content: context.locals.starlightRoute.entry.data?.fcFrame ?? "vNext" },
  });
});
```

You can of course use any of the [route data properties](https://starlight.astro.build/reference/route-data/) in your middleware e.g. to compute a dynamic Open Graph image URL based on the current page's [ID](https://starlight.astro.build/reference/route-data/#id) for example or the [`entry`](https://starlight.astro.build/reference/route-data/#entry) in this example which would contain the `fcFrame` frontmatter value.
